### PR TITLE
Adds support for redirected URLs

### PR DESCRIPTION
**Pull request by bikeath1337**
This pull request adds support for redirected URLs. It enables the maximum redirect retries to be set and adds specs to test the new features.

The default retry attempts is 10. To change, use `FFMPEG.max_http_redirect_attempts`.

### CODE REVIEW

File: lib/ffmpeg/movie.rb [hunk 1]
```diff
@@ -217,14 +217,29 @@ def fix_encoding(output)
...
-      http.request_head(url.path)
-    rescue SocketError, Errno::ECONNREFUSED
+      response = http.request_head(url.path)
+
+      case response
+        when Net::HTTPRedirection then
+          raise FFMPEG::HTTPTooManyRequests if limit == 0
+
+          new_uri = URI(response['Location'])
+          if new_uri.relative?
```
Comment (gmcnaughton) [inline lib/ffmpeg/movie.rb]: `URI#+` will actually do this for you! `new_uri = url + URI(response['Location'])`  It Does The Right Thing (relative uris append, absolute uris replace):

    URI('http://example.com/path/foo.jpg') + URI('other/bar.jpg')
     => #<URI::HTTP http://example.com/path/other/bar.jpg>

    URI('http://example.com/path/foo.jpg') + URI('https://facebook.com/hello')
     => #<URI::HTTPS https://facebook.com/hello>

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -19,6 +40,7 @@ def tmp_path
 end
 
 def start_web_server
+  WEBrick::HTTPStatus::Redirect
```
Comment (gmcnaughton) [inline spec/spec_helper.rb]: I think this snuck in accidentally 😄

**Comment by gmcnaughton** [comment]
I'm not 100% convinced that streamio-ffmpeg should be in the business of checking URLs, vs. just passing them on to the binary.  It seems like a recipe for whack-a-mole!

Some possible gotcha cases:
* does ffmpeg allow ftp:// urls? (this check may stumble on those, I'm not sure)
* does ffmpeg accept untrusted/self-signed SSL certificates? (this check may raise errors on untrusted certificates, unless we set `http.verify_mode = OpenSSL::SSL::VERIFY_NONE`)

Thanks for jumping on this so quick! 👍

File: lib/ffmpeg/movie.rb [hunk 3]
```diff
@@ -17,7 +17,7 @@ def initialize(path)
 
       if remote?
         @head = head
-        raise Errno::ENOENT, "the URL '#{path}' does not exist" if @head.nil? || @head.code.to_i != 200
+        raise Errno::ENOENT, "the URL '#{path}' does not exist" unless @head.is_a?(Net::HTTPSuccess) || @head.is_a?(Net::HTTPRedirection)
```
Comment (bikeath1337) [inline lib/ffmpeg/movie.rb]: True. That was a relic.

File: lib/ffmpeg/movie.rb [hunk 4]
```diff
@@ -217,14 +217,29 @@ def fix_encoding(output)
...
-      url = URI(@path)
+    def head(location=@path, limit=FFMPEG.max_http_redirect_attempts)
+      url = URI(location)
       return unless url.path
 
       http = Net::HTTP.new(url.host, url.port)
       http.use_ssl = url.port == 443
-      http.request_head(url.path)
-    rescue SocketError, Errno::ECONNREFUSED
+      response = http.request_head(url.path)
```
Comment (bikeath1337) [inline lib/ffmpeg/movie.rb]: Will add a spec for this.

File: lib/ffmpeg/movie.rb [hunk 5]
```diff
@@ -217,14 +217,29 @@ def fix_encoding(output)
...
+
+      case response
+        when Net::HTTPRedirection then
+          raise FFMPEG::HTTPTooManyRequests if limit == 0
```
Comment (bikeath1337) [inline lib/ffmpeg/movie.rb]: agree.

File: lib/ffmpeg/movie.rb (hunk 1 above)
Comment (bikeath1337) [inline lib/ffmpeg/movie.rb]: Nice improvement.

**Comment by bikeath1337** [comment]
Thanks for the quick review. I was thinking the same thing as you but went ahead with the work to see where it went because I didn't feel that simply ignoring the url was any better.

File: lib/ffmpeg/movie.rb (hunk 5 above)
Comment (bikeath1337) [inline lib/ffmpeg/movie.rb]: I'm opting to validate the setting, following the pattern already set: >= 0 and a whole number.

**Comment by bikeath1337** [comment]
Branch updated.

**Pull request merged**